File: packages/kotlin/cnos-kotlin-azure/src/main/kotlin/ai/kitsy/cnos/vault/azure/AzureKeyVaultProvider.kt
package ai.kitsy.cnos.vault.azure

import ai.kitsy.cnos.SecretVaultProvider
import ai.kitsy.cnos.SecretVaultProviderFactory
import ai.kitsy.cnos.VaultAuthConfig
import ai.kitsy.cnos.VaultDefinition
import com.azure.identity.DefaultAzureCredentialBuilder
import com.azure.security.keyvault.secrets.SecretClient
import com.azure.security.keyvault.secrets.SecretClientBuilder

class AzureKeyVaultProvider(private val vaultId: String, private val def: VaultDefinition) : SecretVaultProvider {
    private var client: SecretClient? = null

    override fun authenticate(auth: VaultAuthConfig) {
        val vaultUrl = def.mapping["vaultUrl"]
            ?: System.getenv("AZURE_KEYVAULT_URL")
            ?: "https://$vaultId.vault.azure.net"
        client = SecretClientBuilder()
            .vaultUrl(vaultUrl)
            .credential(DefaultAzureCredentialBuilder().build())
            .buildClient()
    }

    override fun batchGet(refs: List<String>): Map<String, Any> {
        val c = client ?: return emptyMap()
        val out = mutableMapOf<String, Any>()
        for (ref in refs) {
            try {
                // Azure Key Vault does not allow dots in secret names
                val name = ref.replace('.', '-')
                val secret = c.getSecret(name)
                out[ref] = secret.value
            } catch (_: Exception) {}
        }
        return out
    }

    override fun get(ref: String): Any? = batchGet(listOf(ref))[ref]

    companion object {
        fun factory(): SecretVaultProviderFactory = SecretVaultProviderFactory("azure") { vaultId, def ->
            AzureKeyVaultProvider(vaultId, def)
        }
    }
}

File: packages/csharp/Kitsy.Cnos.Azure/AzureKeyVaultProvider.cs
using System;
using System.Collections.Generic;
using Azure;
using Azure.Identity;
using Azure.Security.KeyVault.Secrets;
using Kitsy.Cnos;

namespace Kitsy.Cnos.Azure
{
    /// <summary>
    /// CNOS vault provider for Azure Key Vault.
    /// Register via <c>new SecretVaultProviderFactory("azure", (id, def) => new AzureKeyVaultProvider(vaultUrl))</c>.
    /// </summary>
    public sealed class AzureKeyVaultProvider : ISecretVaultProvider
    {
        private readonly string _vaultUrl;
        private SecretClient? _client;

        /// <param name="vaultUrl">Azure Key Vault URL (e.g. "https://myvault.vault.azure.net/").</param>
        public AzureKeyVaultProvider(string vaultUrl)
        {
            _vaultUrl = vaultUrl ?? throw new ArgumentNullException(nameof(vaultUrl));
        }

        public void Authenticate(VaultAuthConfig auth)
        {
            string? url = auth.Config?.TryGetValue("vaultUrl", out object? v) == true ? v?.ToString() : null;
            url ??= _vaultUrl;

            string? token = auth.Token;
            global::Azure.Core.TokenCredential credential = !string.IsNullOrEmpty(token)
                ? new ClientSecretCredential(
                    auth.Config?.TryGetValue("tenantId", out object? t) == true ? t?.ToString() ?? "" : "",
                    auth.Config?.TryGetValue("clientId", out object? c) == true ? c?.ToString() ?? "" : "",
                    token!)
                : new DefaultAzureCredential();

            _client = new SecretClient(new Uri(url), credential);
        }

        public Dictionary<string, object?> BatchGet(IReadOnlyList<string> refs)
        {
            var result = new Dictionary<string, object?>(StringComparer.Ordinal);
            foreach (string @ref in refs)
            {
                object? value = GetSingle(@ref);
                if (value != null) result[@ref] = value;
            }
            return result;
        }

        public object? Get(string reference) => GetSingle(reference);

        private object? GetSingle(string secretName)
        {
            if (_client == null)
                throw new CnosError("cnos: AzureKeyVaultProvider.Authenticate must be called before reading secrets");
            try
            {
                Response<KeyVaultSecret> response = _client.GetSecret(secretName);
                return response.Value.Value;
            }
            catch (RequestFailedException ex) when (ex.Status == 404)
            {
                return null;
            }
            catch (Exception ex) when (ex is not CnosError)
            {
                throw new CnosError($"cnos: Azure Key Vault error for \"{secretName}\": {ex.Message}", ex);
            }
        }

        /// <summary>Returns a <see cref="SecretVaultProviderFactory"/> for this provider.</summary>
        public static SecretVaultProviderFactory Factory(string vaultUrl) =>
            new SecretVaultProviderFactory("azure", (_, _) => new AzureKeyVaultProvider(vaultUrl));
    }
}
